drivers/hmis/app/graphql/types/hmis_schema/query_type.rb

```diff
@@ -554,13 +554,18 @@ def ce_opportunity(id:)
     def ce_referral_step(id:)
       raise unless Hmis::Ce.configuration.enabled?
 
-      Hmis::WorkflowExecution::Step.viewable_by(current_user).find(id)
+      step = Hmis::WorkflowExecution::Step.find(id)
+
+      # Referral permission governs step viewing permission. If you can view a referral, you can view all its steps.
+      referral = Hmis::Ce::Referral.viewable_by(current_user).find_by(workflow_instance: step.instance)
+      access_denied! unless referral
+
+      step
     end
 
-    # Globally available CE opportunities gated by admin permission
+    # All CE opportunities the user can view, resolved on admin page
     ce_opportunities_field
     def ce_opportunities(**args)
-      # TODO(#7506) - gated by admin permission
```

Comment:
Is there a permission that gates visibility of the "admin ce" pages? (viewing all available oppts and all referrals)

Reply:
Thanks, yes, good question that I was also thinking about as I put up this PR. I'll start a comment thread over in the permissions grid document!